pkg/provisioners/gcppubsub/dispatcher/dispatcher/reconcile.go

```diff
@@ -0,0 +1,331 @@
+	}
+}
+
+// subscriptionKey creates the second index into reconciler.subscriptions, based on the Subscriber's
+// name.
+func subscriptionKey(sub *v1alpha1.ChannelSubscriberSpec) subscriptionName {
+	return types.NamespacedName{
+		Namespace: sub.Ref.Namespace,
+		Name:      sub.Ref.Name,
+	}
+}
+
+// stopAllSubscriptions stops listening to all GCP PubSub Subscriptions for the given Channel.
+func (r *reconciler) stopAllSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+	r.stopAllSubscriptionsUnderLock(ctx, c)
+}
+
+// stopAllSubscriptionsUnderLock stops listening to all GCP PubSub Subscriptions for the given
+// Channel.
+// Note that it can only be called if reconciler.subscriptionsLock is held.
+func (r *reconciler) stopAllSubscriptionsUnderLock(ctx context.Context, c *eventingv1alpha1.Channel) {
+	channelKey := key(c)
+	if subscribers, present := r.subscriptions[channelKey]; present {
+		for _, subCancel := range subscribers {
+			subCancel()
+		}
+	}
+	delete(r.subscriptions, channelKey)
+}
+
+// syncSubscriptions ensures all subscribers of the Channel have a background Goroutine that is
+// polling the GCP PubSub Subscriptions representing it. It also removes listeners from Subscribers
+// that no longer exist.
+func (r *reconciler) syncSubscriptions(ctx context.Context, c *eventingv1alpha1.Channel) error {
+	r.subscriptionsLock.Lock()
+	defer r.subscriptionsLock.Unlock()
+
+	// We are going to manipulate subscribers, but don't want it written back, so make a throw away
+	// copy.
+	subscribers := c.Spec.Subscribable.DeepCopy()
+	if subscribers == nil {
+		// There are no subscribers.
+		r.stopAllSubscriptionsUnderLock(ctx, c)
+		return nil
+	}
+
+	for _, subscriber := range subscribers.Subscribers {
+		err := r.createSubscriptionUnderLock(loggingWith(ctx, zap.Any("subscriber", subscriber)), c, &subscriber)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Now remove all subscriptions that are no longer present.
+	channelKey := key(c)
+	activeSubscribers := r.subscriptions[channelKey]
+	if len(subscribers.Subscribers) == len(activeSubscribers) {
+		return nil
+	}
+
+	// subsToDelete is logically a set, not a map (values have no meaning).
+	subsToDelete := map[subscriptionName]bool{}
+	for sub := range activeSubscribers {
+		subsToDelete[sub] = true
+	}
+	for _, sub := range subscribers.Subscribers {
+		delete(subsToDelete, subscriptionKey(&sub))
+	}
+	for subToDelete := range subsToDelete {
+		r.subscriptions[channelKey][subToDelete]()
+	}
+	return nil
+}
+
+// createSubscriptionUnderLock starts a background Goroutine for a single subscriber polling its
+// GCP PubSub Subscription.
+// Note that it can only be called if reconciler.subscriptionsLock is held.
+func (r *reconciler) createSubscriptionUnderLock(ctx context.Context, c *eventingv1alpha1.Channel, sub *v1alpha1.ChannelSubscriberSpec) error {
+	ctxWithCancel, cancelFunc := context.WithCancel(ctx)
```

Comment:
nit: This can be moved after the check at [line 255](https://github.com/knative/eventing/pull/618/files#diff-45dd13866950a11f956d59cf453929ceR255) to save a context allocation.

Reply:
Done.